rfc/manifest-validation.md

```diff
@@ -0,0 +1,451 @@
+   */
+  validate(manifests: string[]): Violation[];
+
+}
+
+Where the Violation struct is defined as follows:
+
+/**
+ * Represents a distinct violation in the manifest.
+ */
+export interface Violation {
+  /**
+   * The name of resource exhibiting the violation.
+   */
+  readonly resourceName: string;
+  /**
+   * The violation message.
+   */
+  readonly message: string;
+
+  /**
+   * The manifest file this violation originated in.
+   */
+  readonly manifestPath: string;
+}
+```
+
+With these interfaces in-place, this is how a datree implementation would potentially look like:
+
+```ts
+import { Validation, Violation } from 'cdk8s-cli';
+
+export interface DatreeValidationProps {
+  /**
+   * URL for the policy to use.
+   */
+  readonly policyUrl: string;
+}
+
+export class DatreeValidation implements Validation {
+
+  private readonly policyUrl: string;
+
+  constructor(props: DatreeValidationProps) {
+    this.policyUrl = props.policyUrl;
+  }
+
+  public validate(manifests: string[]): Violation[] {
+    const report: Violation[] = [];
+    for (const manifest of manifests) {
+      const out = execSync(`datree test --output json ${manifest}`);
+      report.push(...parseOutput(out));
+    }
+    return report;
+  }
+}
+```
+
+During cdk8s synth, the CLI will look at the validations config property, download it from
+the remote source if necessary, and proceed to load and execute the plugins.
+
+Optionally, we could also provide a cdk8s validate command that can operate independently from synthesis.
+Instead of synthesizing manifests, it can accept them as a command line argument. This command can then
+be used in processes that don’t have access to the original source code, and operate solely on the output
+of cdk8s synth. One such example can be the deploy time validation gate we mentioned in the beginning.
+
+_FAQ_
+
+*Q*. Who is responsible for datree CLI to exist on the system? the user? the plugin? cdk8s?
+*A. *This would be for the plugin to decide. If embedding the tool inside an npm package is viable,
+they can do that. If not, the plugin will probably need to provide installation instructions.
+
+*Q*. Should cdk8s synth fail if the validation results in violations?
+*A.* Intuitively I’d say yes, but in order to debug the violation, the manifest itself will most likely be needed.
+So we will need to generate it, even if the validation fails. We should probably still exist with non-zero though, to indicate an error.
+
+### Construct Aware
+
+The ability to traceback to constructs world, will be enabled by adding a few annotations to the manifests generated
+by cdk8s. More concretely, we add cdk8s.io/construct.path and cdk8s.io/construct.traceback annotations:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: chart-deployment-c8a3b439
+  annotations:
+    cdk8s.io/construct.path: Chart/Deployment
+    cdk8s.io/construct.traceback:
+      - deployment.ts:25
+      - main.ts:15
```

Comment:
I like this! 👍  It's a little bit like a source map, but just with the information needed for the given resource.